### PR TITLE
Expose RoundRobinInetAddressResolver

### PR DESCRIPTION
**Pull request by Spikhalskiy**
Motivation:
Make small refactoring for recently merged PR #5867 to make the code more flexible and expose aggressive round robin as a `NameResolver` too with proper code reuse.

Modifications:
Round robin is a method of hostname resolving - so Round robin related code fully moved to `RoundRobinInetAddressResolver implements NameResolver<InetAddress>`, `RoundRobinInetSocketAddressResolver` is deleted as a separate class, an instance with the same functionality could be created by calling `#asAddressResolver()`.

Result:
New forced Round Robin code exposed not only as an `AddressResolver` but as a `NameResolver` too, more proper code and semantic reusing of `InetNameResolver` and `InetSocketAddressResolver` classes.

### CODE REVIEW

File: resolver/src/main/java/io/netty/resolver/RoundRobinInetAddressResolver.java [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+
+/**
+ * A {@link NameResolver} that resolves {@link InetAddress} and force Round Robin by choosing a single address
+ * randomly in {@link #resolve(String)} and {@link #resolve(String, Promise)}
+ * if multiple are returned by the {@link NameResolver}.
+ * Use {@link #asAddressResolver()} to create a {@link java.net.InetSocketAddress} resolver
+ */
+@UnstableApi
+public class RoundRobinInetAddressResolver extends InetNameResolver {
+    private NameResolver<InetAddress> nameResolver;
```
Comment (normanmaurer) [inline resolver/src/main/java/io/netty/resolver/RoundRobinInetAddressResolver.java]: final

File: resolver/src/main/java/io/netty/resolver/RoundRobinInetAddressResolver.java [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+                            List<InetAddress> inetAddresses = future.getNow();
+                            int numAddresses = inetAddresses.size();
+                            if (numAddresses == 0) {
+                                promise.setFailure(new UnknownHostException(inetHost));
+                            } else {
+                                // if there are multiple addresses: we shall pick one at random
+                                // this is to support the round robin distribution
+                                int index =
+                                        (numAddresses == 1)? 0 : ThreadLocalRandom.current().nextInt(numAddresses);
+                                promise.setSuccess(inetAddresses.get(index));
```
Comment (normanmaurer) [inline resolver/src/main/java/io/netty/resolver/RoundRobinInetAddressResolver.java]: just inline the index calculation ?

File: resolver/src/main/java/io/netty/resolver/RoundRobinInetAddressResolver.java [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+
+import java.net.InetAddress;
+import java.net.UnknownHostException;
+import java.util.List;
+
+/**
+ * A {@link NameResolver} that resolves {@link InetAddress} and force Round Robin by choosing a single address
+ * randomly in {@link #resolve(String)} and {@link #resolve(String, Promise)}
+ * if multiple are returned by the {@link NameResolver}.
+ * Use {@link #asAddressResolver()} to create a {@link java.net.InetSocketAddress} resolver
```
Comment (normanmaurer) [inline resolver/src/main/java/io/netty/resolver/RoundRobinInetAddressResolver.java]: Import `java.net.InetSocketAddress` and so not need to use FQCN here.

File: resolver/src/main/java/io/netty/resolver/RoundRobinInetAddressResolver.java [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+     * @param nameResolver the {@link NameResolver} used for name resolution
+     */
+    public RoundRobinInetAddressResolver(EventExecutor executor, NameResolver<InetAddress> nameResolver) {
+        super(executor);
+        this.nameResolver = nameResolver;
+    }
+
+    @Override
+    protected void doResolve(final String inetHost, final Promise<InetAddress> promise) throws Exception {
+        // hijack the doResolve request, but do a doResolveAll request under the hood
```
Comment (normanmaurer) [inline resolver/src/main/java/io/netty/resolver/RoundRobinInetAddressResolver.java]: missing `.` on EOL.

File: resolver/src/main/java/io/netty/resolver/RoundRobinInetAddressResolver.java [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+        // Note that InetSocketAddress.getHostName() will never incur a reverse lookup here,
+        // because an unresolved address always has a host name.
+        resolveAll(inetHost)
+                .addListener(new FutureListener<List<InetAddress>>() {
```
Comment (normanmaurer) [inline resolver/src/main/java/io/netty/resolver/RoundRobinInetAddressResolver.java]: merge both above lines

**Comment by Spikhalskiy** [comment]
@normanmaurer Addressed your comments. I'm thinking about rework ```ThreadLocalRandom.current().nextInt(numAddresses)``` to just stupid ```private int index = 0; inetAddresses.get(index++/inetAddresses.size())```, even if it's a bit not thread-safe. We don't actually need an expensive random call here. WDYT?

**Comment by normanmaurer** [comment]
@Spikhalskiy let us look into this after this is merged..

**Comment by normanmaurer** [comment]
Cherry-picked into 4.1 (eb7f8e4dc5802fa7ec1ffbfbc8ac8ef1aacf8918)

**Pull request closed without merging**